src/main/java/com/google/devtools/build/lib/remote/Scrubber.java

```diff
@@ -0,0 +1,156 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.base.Predicates;
+import com.google.common.collect.ImmutableMap;
+import com.google.devtools.build.lib.actions.ActionInput;
+import com.google.devtools.build.lib.actions.Spawn;
+import com.google.devtools.build.lib.actions.cache.VirtualActionInput;
+import com.google.devtools.build.lib.remote.options.RemoteOptions;
+import com.google.devtools.common.options.RegexPatternOption;
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * The {@link Scrubber} implements scrubbing of remote cache keys.
+ *
+ * <p>See the documentation for the {@code --experimental_remote_scrub_*} flags for more
+ * information.
+ */
+public class Scrubber {
+
+  /**
+   * A {@link SpawnScrubber} determines how to scrub the cache key for a {@link Spawn}.
+   */
+  public static class SpawnScrubber {
+
+    private final Predicate<ActionInput> inputMatcher;
+    private final ImmutableMap<RegexPatternOption, String> argReplacements;
+    private final String salt;
+
+    private SpawnScrubber(RemoteOptions options) {
+      this.inputMatcher = buildInputMatcher(options);
+      this.argReplacements = ImmutableMap.copyOf(options.scrubArgReplacements);
+      this.salt = options.scrubSalt;
+    }
+
+    private static Predicate<ActionInput> buildInputMatcher(RemoteOptions options) {
+      Predicate<String> execPathMatcher = buildStringMatcher(options.scrubInput);
+      return (input) -> !input.equals(VirtualActionInput.EMPTY_MARKER) && execPathMatcher.test(input.getExecPathString());
+    }
+
+    /**
+     * Whether the given input should be omitted from the cache key.
+     */
+    public boolean shouldOmitInput(ActionInput input) {
+      return inputMatcher.test(input);
+    }
+
+    /**
+     * Transforms an action command line argument.
+     */
+    public String transformArgument(String arg) {
+      for (Map.Entry<RegexPatternOption, String> entry : argReplacements.entrySet()) {
+        Pattern pattern = entry.getKey().regexPattern();
+        String replacement = entry.getValue();
+        // Don't use Pattern#replaceFirst because it allows references to capture groups.
+        Matcher m = pattern.matcher(arg);
+        if (m.find()) {
+          arg = arg.substring(0, m.start()) + replacement + arg.substring(m.end());
+        }
+      }
+      return arg;
+    }
+
+    /**
+     * Returns the scrubbing salt.
+     */
+    public String getSalt() {
+      return salt;
```

Comment:
I don't believe that's possible without the toolchains providing additional information (e.g., a logical "version number") which is out of scope for this PR.

It's possible for an input path, or command line argument, to only exist in one of the platforms that are sharing a cache. So the hypothetical inferred salt for platform 1 can't capture the fact that we omitted that input path, or transformed that argument, because in platform 2 we might not even know that we would have scrubbed it.

Reply:
Thinking some more about it: another idea is to default to using a hash of the `Transform` proto that was used to do the scrubbing. I worry that it would make it too magical, though (and you can definitely do things that would affect the hash, like reordering repeated fields, that might not affect the final result).

@oliviernotteghem, thoughts? Would you prefer Bazel to intelligently compute a suitable salt from the configuration, or to leave it entirely in the user's hands?

Reply:
I'd rather leave it to the user's hands. Re-hashing everything can be costly for CI/engs on a large codebase, so avoiding it is a nice to have. I like the determinism too of having it manual. Anecdotically, we've has to bump that salt recently to fix an (unrelated) cache poisining issue, so it can be handy for other use cases as well ;D

Reply:
Leaving as-is for now.